lib/orbit-common/cache/observables/has-one-observable.js

```diff
@@ -0,0 +1,78 @@
+import { Observable } from 'rxjs/Observable';
+import 'rxjs/add/operator/filter';
+import 'rxjs/add/operator/map';
+import 'rxjs/add/operator/do';
+import 'rxjs/add/operator/last';
+import 'rxjs/add/operator/scan';
+import 'rxjs/add/operator/startWith';
+import 'rxjs/add/observable/concat';
+import 'rxjs/add/observable/empty';
+import 'orbit-common/rxjs/operators/matching';
+import RecordsObservable from './records-observable';
+
+function extractRecordFromHasOneResult(value) {
+  if (!value) { return null; }
+
+  const id = Object.keys(value)[0];
+  return value[id];
+}
+
+export default class HasOneObservable extends Observable {
+  constructor(subscribe, cache, record, relationship) {
+    super(subscribe);
+    this.cache = cache;
+    this.record = record;
+    this.relationship = relationship;
+  }
+
+  lift(operator) {
+    const observable = new HasOneObservable();
+    observable.source = this;
+    observable.operator = operator;
+    observable.cache = this.cache;
+    observable.record = this.record;
+    observable.relationship = this.relationship;
+    return observable;
+  }
+
```

Comment:
This is a little more fiddly than it needs to be because of the mismatch between `replaceHasOne` and the operations that are produced here, `addRecord` and `removeRecord`. When `replaceHasOne` with `null` is received it doesn't know which record has been replaced and can't find out from the cache because the reference has already gone. For this reason `scan` is used to memoize the previous value (`scan` works the same way as `reduce` but over Observables rather than Arrays).